include/unifex/detail/concept_macros.hpp

```diff
@@ -0,0 +1,489 @@
+    ::unifex::requires_<__VA_ARGS__>
+  #define UNIFEX_CONCEPT_FRAGMENT_REQS_REQUIRES_typename(...) \
+    static_cast<::unifex::_concept::tag<__VA_ARGS__> *>(nullptr)
+  #if defined(__GNUC__) && !defined(__clang__)
+    // GCC can't mangle noexcept expressions, so just check that the
+    // expression is well-formed.
+    // https://gcc.gnu.org/bugzilla/show_bug.cgi?id=70790
+    #define UNIFEX_CONCEPT_FRAGMENT_REQS_REQUIRES_noexcept(...) \
+      __VA_ARGS__
+  #else
+    #define UNIFEX_CONCEPT_FRAGMENT_REQS_REQUIRES_noexcept(...) \
+      ::unifex::requires_<noexcept(__VA_ARGS__)>
+  #endif
+
+  #define UNIFEX_FRAGMENT(NAME, ...) \
+    (1u==sizeof(NAME ## UNIFEX_CONCEPT_FRAGMENT_( \
+      static_cast<::unifex::_concept::tag<__VA_ARGS__> *>(nullptr), nullptr)))
+
+  #define UNIFEX_TYPE(...) \
+    ::unifex::_concept::first_t<__VA_ARGS__, decltype(UNIFEX_arg)> \
+    /**/
+  #define UNIFEX_AS_TYPE_(ARG) \
+    , UNIFEX_TYPE(UNIFEX_PP_IIF(UNIFEX_PP_NOT(UNIFEX_PP_IS_PAREN(ARG))) \
+        (, CPP_PP_EXPAND) ARG) \
+    /**/
+  #define UNIFEX_DEFER_(CONCEPT, ...) \
+    true \
+      ? nullptr \
+      : ::unifex::_concept::make_boolean( \
+        [](auto UNIFEX_arg) { \
+          (void) UNIFEX_arg; \
+          return std::integral_constant<bool, (CONCEPT<__VA_ARGS__>)>{}; \
+        }) \
+    /**/
+  #define UNIFEX_DEFER(CONCEPT, ...) \
+    UNIFEX_DEFER_( \
+      CONCEPT, \
+      UNIFEX_PP_EVAL( \
+        UNIFEX_PP_TAIL, \
+        ~ UNIFEX_PP_FOR_EACH(UNIFEX_AS_TYPE_, __VA_ARGS__))) \
+    /**/
+
+#endif
+
+#ifdef UNIFEX_WORKAROUND_MSVC_654601
+  #define UNIFEX_FORCE_TO_BOOL static_cast<bool>
+#else
+  #define UNIFEX_FORCE_TO_BOOL
+#endif
+
+////////////////////////////////////////////////////////////////////////////////
+// UNIFEX_TEMPLATE
+// Usage:
+//   UNIFEX_TEMPLATE(typename A, typename B)
+//     (requires Concept1<A> && Concept2<B>)
+//   void foo(A a, B b)
+//   {}
+#if UNIFEX_CXX_CONCEPTS
+  #define UNIFEX_TEMPLATE(...) \
+    template<__VA_ARGS__> UNIFEX_PP_EXPAND \
+    /**/
+  #define UNIFEX_TEMPLATE_DEF UNIFEX_TEMPLATE \
+    /**/
+#else
+  #define UNIFEX_TEMPLATE \
+    UNIFEX_TEMPLATE_SFINAE \
+    /**/
+  #define UNIFEX_TEMPLATE_DEF UNIFEX_TEMPLATE_DEF_SFINAE \
+    /**/
+#endif
+
+#define UNIFEX_TEMPLATE_SFINAE(...) \
+  template<__VA_ARGS__ UNIFEX_TEMPLATE_SFINAE_AUX_ \
+  /**/
+#define UNIFEX_TEMPLATE_SFINAE_AUX_(...) , \
+  typename UNIFEX_true_ = std::true_type, \
+  std::enable_if_t< \
+    UNIFEX_FORCE_TO_BOOL( \
+      UNIFEX_PP_CAT(UNIFEX_TEMPLATE_SFINAE_AUX_3_, __VA_ARGS__) && UNIFEX_true_{}), \
+    int> = 0> \
+  /**/
+#define UNIFEX_TEMPLATE_DEF_SFINAE(...) \
+  template<__VA_ARGS__ UNIFEX_TEMPLATE_DEF_SFINAE_AUX_ \
+  /**/
+#define UNIFEX_TEMPLATE_DEF_SFINAE_AUX_(...) , \
+  typename UNIFEX_true_, \
+  std::enable_if_t< \
+    UNIFEX_FORCE_TO_BOOL( \
+      UNIFEX_PP_CAT(UNIFEX_TEMPLATE_SFINAE_AUX_3_, __VA_ARGS__) && UNIFEX_true_{}), \
+    int>> \
```

Comment:
should the macros use the `int(*)[__LINE__]` pattern to disambiguate overloads?

Reply:
It hasn't been necessary, and I'm not sure why. I'll leave it off until we need it.